File: 2020/python/day3.py
from std import *

data = []
for line in lines("../input/input3.txt"):
    data.append(line)


def slope(dx, dy):
    x = 0
    tree = 0
    for y in range(0, len(data), dy):
        if data[y][x] == '#':
            tree += 1
        x = (x + dx) % len(data[y])

    return tree    

def partA():
    return slope(3,1)

def partB():
    return slope(1,1)*slope(3, 1)*slope(5, 1)*slope(7, 1)*slope(1, 2)

print("A:", partA())
print("B:", partB())

File: 2020/haskell/day3.hs
ski [] _ _ = 0
ski (f:fs) x delta@(dx, dy) = cnt + ski nextY nextX delta
        where
                cnt = if f!!x == '#' then 1 else 0      -- do the tree counting
                nextX = mod (x+dx) (length f)           -- move x-wise
                nextY = drop (dy-1) fs                  -- move y-wise

main = do 
        c <- readFile "../input/input3.txt"
        let field = lines c
        let slopes = map (ski field 0) [(3,1), (1,1), (5,1), (7,1), (1,2)]
        putStrLn ("A: " ++ show (head slopes))
        putStrLn ("B: " ++ show (product slopes))